### PR TITLE
feat(rds): simpler way to configure parameters for instance and cluster

### LINKED ISSUE #18124: (rds): simpler way to configure parameters for instance or cluster
### Description

Instead of creating a new ParameterGroup and referencing it in the props, add a new field `readonly parameters?: { [key: string]: string };` to the instance and cluster props that is used to create a ParameterGroup under the hood.

### Use Case

Easier way to configure parameters. The parameters field is the only field that is necessary to create a group so it could be provided directly instead of forcing the user to create a group.

### Proposed Solution

Add `readonly parameters?: { [key: string]: string }; ` to DatabaseInstanceNew properties and create ParameterGroup internally.

### Other information

_No response_

### Acknowledge

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by hoegertn**
Fixes #18124

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:834]: Curious why are we doing this in `DatabaseInstanceSourceProps`, and not in `DatabaseInstanceNewProps`, which is where `parameterGroup` is defined?

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:911]: Same question here - I wonder why is this done here in `DatabaseInstanceSource`, and not in `DatabaseInstanceNew`, which handles `dbParameterGroupName` of the L1?

File: packages/@aws-cdk/aws-rds/test/instance.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:14]: Nope! Use the qualified import (`rds`) from above please 🙂.

File: packages/@aws-cdk/aws-rds/test/instance.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:278]: No need for this empty line.

**Review by skinny85 — COMMENTED** [review]
Looks good @hoegertn! A few questions.

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:834]: The DatbaseInstanceNew does not know the engine yet, so we cannot create a ParameterGroup.

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:911]: see above

**Comment by hoegertn** [comment]
I also added the cluster now. Currently, there are no README changes as the parameter group is also not documented. How should I proceed here?

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (jogold) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:359]: I would let the default of `ParameterGroup` act here.

```suggestion
```

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (jogold) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:757]:
```suggestion
```

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (jogold) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:908]:
```suggestion
```

**Comment by skinny85** [comment]
> I also added the cluster now. Currently, there are no README changes as the parameter group is also not documented. How should I proceed here?

Let's add a section in the ReadMe, showing off the two ways of setting a `ParameterGroup` for an Instance/Cluster, and when would you use which one.

**Comment by hoegertn** [comment]
@skinny85 Is there anything missing from my side?

**Comment by skinny85** [comment]
> @skinny85 Is there anything missing from my side?

Probably nothing beyond the fact that you didn't re-request my approval, so I missed this 😃.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts:359]: Yep, +1.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:362]: Formatting seems a little off here:

```suggestion
      props.parameters
        ? new ParameterGroup(this, 'ParameterGroup', {
          engine: props.engine,
          parameters: props.parameters,
        })
        : undefined
```

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:353]:
```suggestion
      throw new Error('You cannot specify both parameterGroup and parameters');
```

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts:757]: +1.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:750]:
```suggestion
    throw new Error('You cannot specify both parameterGroup and parameters');
```

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:760]:
```suggestion
    instanceProps.parameters
      ? new ParameterGroup(cluster, 'InstanceParameterGroup', {
        engine: props.engine,
        parameters: instanceProps.parameters,
      })
      : undefined
```

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:834]: Makes sense.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:749]: Should this not be

```suggestion
  if (instanceProps.parameterGroup && instanceProps.parameters) {
```

?

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:908]: +1.

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:911]: Yep.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:341]: I think we can get rid of this.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:375]: Missing a negative test case (when both `parameterGroup` and `parameters` were provided, for both the Cluster, and its Instances). I think that's why the above typo snuck through.

File: packages/@aws-cdk/aws-rds/test/instance.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:278]: Missing negative test (when both `parameters` and `parameterGroup` was provided).

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks good @hoegertn! A few minor comments, mainly around docs.

File: packages/@aws-cdk/aws-rds/README.md
Comment (hoegertn) [reply packages/@aws-cdk/aws-rds/README.md:566]: Unversioned engines are not supported by ParameterGroups afaik

**Comment by hoegertn** [comment]
Sorry, had some busy weeks. I need to write the missing tests and then this is ready. I plan to finish this within the next days.

**Comment by skinny85** [comment]
Sounds great! No rush, I just wanted to make sure you weren't waiting on my review 🙂.

Re-request my review (there's a button in the top-right of the PR window, next to my avatar) once you're ready for another round of reviews!

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:910]: Can we extract this all of this to a `parameterGroupName: string | undefined` local variable? This code is way, way too clever for my liking.

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:830]:
```suggestion
   * You need to use a versioned engine to auto-generate a DBParameterGroup.
```

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:361]:
```suggestion
      props.parameters
        ? new ParameterGroup(this, 'ParameterGroup', {
          engine: props.engine,
          parameters: props.parameters,
        })
        : undefined
```

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:758]:
```suggestion
    instanceProps.parameters
      ? new ParameterGroup(cluster, 'InstanceParameterGroup', {
        engine: props.engine,
        parameters: instanceProps.parameters,
      })
      : undefined
```

File: packages/@aws-cdk/aws-rds/README.md
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/README.md:559]: Let's use a colon, like we do above:

```suggestion
You can use this if you do not want to reuse the parameter group instance for different instances:
```

File: packages/@aws-cdk/aws-rds/lib/props.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/props.ts:50]:
```suggestion
   * You need to use a versioned engine to auto-generate a DBParameterGroup.
```

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/test/cluster.test.ts:341]: Still not needed.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:333]: Doesn't this test contradict this statement from the docs?

>    * You need to use a versioned engine to auto-generate a DBParameterGroup

Do the docs need updating?

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:336]: This is optional, right? I would remove it from this test.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:338]: Can we make it something silly, like `'cluster_locks'` and `'instance_locks'`, instead of `100` and `200`, which are a little more tricky to remember which is which in each test?

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:384]: This is optional, I would remove it from this test.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:401]: This is optional, I would remove it from this test.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:395]: This is optional, I would remove it from this test.

File: packages/@aws-cdk/aws-rds/test/instance.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:253]: This is optional, I would remove it from this test.

File: packages/@aws-cdk/aws-rds/test/instance.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:279]: This is optional, I would remove it from this test.

File: packages/@aws-cdk/aws-rds/test/instance.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/instance.test.ts:288]: This is optional, I would remove it from this test.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:410]: I believe we are still missing a test where `parameters` and `parameterGroup` where specified both in `InstanceProps` of the Cluster.

**Review by skinny85 — DISMISSED** [review]
Looks great @hoegertn! Minor stylistic comments before we merge this in.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-rds/test/cluster.test.ts:333]: Seems that Aurora has an internal default version other engines do not have.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-rds/test/cluster.test.ts:335]: At least now it is a valid value

**Review by skinny85 — APPROVED** [review]
👍

**Pull request merged**
